### DISCUSSION TITLE
Restore using nuget.exe based on a msBuild Condition

### DISCUSSION THREAD

**Discussion by StarWars999123** [NuGet/Home#11452] [Q&A] [answered]
I got a project with multiple SDKs and due to different build solutions, this project gets compiled on different agents. Some with older Visual Studio versions, not supporting NET6 yet. 
As msbuild supports the condition attribute, I applied this switch here as described here: https://stackoverflow.com/a/53539758/4169550
```
<TargetFrameworks Condition="'$(BuildNetStdOnly)' == 'true'">netstandard2.0</TargetFrameworks>
<TargetFrameworks Condition="'$(BuildNetFxOnly)' == 'true'">net471</TargetFrameworks>
<TargetFrameworks Condition="'$(TargetFrameworks)' == ''">netstandard2.0;net471</TargetFrameworks>
```
Before the build starts, I try to restore the packages using
```
nuget.exe -restore MySolutionWithoutNet6.sln
```
Depending on the build solution, I either use the first or second msbuild call: 
```
msbuild -p:BuildNetStdOnly=true -p:Configuration=Release -m -nr:false
msbuild -p:BuildNetFxOnly=true -p:Configuration=Release -m -nr:false
```

As the older Visual Studio does not support .NET6, I fails. So, I am missing an switch to set such a variable and allow to restore the solution using the argument switch. 
Do you have any different approach to resolve this issue?

**Comment by zivkan** [comment, marked as answer]
Instead of using nuget.exe to restore, you can use msbuild: https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restore-target

```text
msbuild -t:restore my.sln
```

If you need to set properties on the command line, you already know how to do that with msbuild. If all the projects in the solution are SDK style, you probably don't need MSBuild and can use the dotnet CLI instead: `dotnet restore my.sln`.


For CI agents, have you considered the dotnet-install scripts: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script

Many Microsoft repos use it in their CI builds, including most, if not all of, github.com/dotnet, and NuGet.Client does as well.

**Comment by StarWars999123** [comment]
@zivkan 
Thank you, your anwers were also quite good. I had to stay with the VS task, but found even a further way to fix it: 

- Definition of the environment variable (for TFS/VS pipelines: just define a pipeline variable) NUGET_RESTORE_MSBUILD_ARGS e.g. as "/p:targetFramework=net452" 
- nuget.exe restore as usual 
- msbuild must be called with MsBuildArgs value as set before (e.g.) "/p:targetFramework=net452"

**Reply by zivkan** [reply]
I still highly recommend getting off nuget.exe and use `msbuild -t:restore` instead. With PackageReference, NuGet writes an intermediate file which the rest of the .NET SDK reads. From time to time we need to add more information to the file, and if you restore using an older version of nuget.exe, it will miss that new data and the project/package won't work correctly. We haven't yet added backwards incompatible changes, but if we ever do, then using a newer version of nuget.exe with an older version of the .net sdk or msbuild will also cause problems.

If you use msbuild or dotnet cli restore instead of nuget.exe, then you never have to worry about nuget.exe versions again, as you'll always be using the version of NuGet that ships with msbuild/dotnet.
